gtsfm/loader/argoverse_dataset_loader.py

```diff
@@ -0,0 +1,132 @@
+            log_id: unique ID of vehicle log
+            stride: sampling rate, e.g. every 2 images, every 4 images, etc.
+            max_num_imgs: number of images to load (starting from beginning of log sequence)
+        """
+        self.log_id = log_id
+        self.dl = SimpleArgoverseTrackingDataLoader(data_dir=dataset_dir, labels_dir=dataset_dir)
+
+        max_lookahead_sec = 60  # at original 30 fps frame rate, so 2 sec
+        self.max_lookahead_imgs = max_lookahead_sec / stride
+
+        self.calib_data = self.dl.get_log_calibration_data(log_id)
+        camera_name = "ring_front_center"
+
+        image_paths = self.dl.get_ordered_log_cam_fpaths(log_id, camera_name)
+        image_paths = image_paths[::stride]
+        self.image_paths = image_paths[:max_num_imgs]
+
+        # for each image, cache its associated timestamp
+        self.image_timestamps = [int(Path(path).stem.replace(f"{camera_name}_", "")) for path in self.image_paths]
+
+        cam_config = get_calibration_config(self.calib_data, camera_name)
+        self.K = cam_config.intrinsic[:3, :3]
+
+        self.camera_SE3_egovehicle = SE3(rotation=cam_config.extrinsic[:3, :3], translation=cam_config.extrinsic[:3, 3])
+        self.egovehicle_SE3_camera = self.camera_SE3_egovehicle.inverse()
+
+    def __len__(self) -> int:
+        """
+        The number of images in the dataset.
+
+        Returns:
+        the number of images.
+        """
+        return len(self.image_paths)
+
+    def get_image(self, index: int) -> Image:
+        """
+        Get the image at the given index.
+
+        Args:
+            index: the index to fetch.
+
+        Raises:
+            IndexError: if an out-of-bounds image index is requested.
+
+        Returns:
+            Image: the image at the query index.
+        """
+
+        if index < 0 or index > self.__len__():
+            raise IndexError("Image index is invalid")
+
+        return io_utils.load_image(self.image_paths[index])
+
+    def get_camera_intrinsics(self, index: int) -> Optional[Cal3Bundler]:
+        """Get the camera intrinsics at the given index.
+
+        Args:
+            the index to fetch.
+
+        Returns:
+            intrinsics for the given camera.
+        """
+        return Cal3Bundler(
+            fx=min(self.K[0, 0], self.K[1, 1]),
+            k1=0,
+            k2=0,
+            u0=self.K[0, 2],
+            v0=self.K[1, 2],
+        )
+
+    def get_camera_pose(self, index: int) -> Optional[Pose3]:
+        """Get the camera pose (in world coordinates) at the given index.
+
+        World coordinate frame is the "city" coordinate frame.
+
+        Args:
+            index: the index to fetch.
+
+        Returns:
+            the camera pose w_P_index.
+        """
+        timestamp = self.image_timestamps[index]
+        city_SE3_egovehicle = self.dl.get_city_SE3_egovehicle(self.log_id, timestamp)
+        if city_SE3_egovehicle is None:
+            return None
```

Comment:
Will it be None for all indices or None for select indices?

Reply:
just for a single index

Reply:
This can get tricky here. I have assumed that we either get no pose or all pose from a loader (e.g. while evaluation).

As we are already subsampling, can we do some non-uniform sampling to ensure we always pick images with a camera pose?

Reply:
Good point, sure, I can do that

Reply:
should be fixed now